### PR TITLE
test: create unit tests

### PR DESCRIPTION
**Pull request by Fyko**
**Please describe the changes this PR makes and why it should be merged:**
This PR creates unit tests in all possible circumstances (where sensible).

Old tests were moved into `test/manual`.

Todo:
- [x] Bitfield
- [x] DataResolver
- [x] SnowflakeUtil
- [x] Util

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: test/Util.test.js [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+      expect(Util.escapeMarkdown(input)).toStrictEqual('\\~\\~hello\\~\\~');
+    });
+
+    test('escape spoiled text', () => {
+      const input = '||hello||';
+      expect(Util.escapeMarkdown(input)).toStrictEqual('\\|\\|hello\\|\\|');
+    });
+  });
+
+  describe('Util#parseEmoji', () => {
```
Comment (Fyko) [inline test/Util.test.js:106]: When would an emoji ever be URI encoded?
https://github.com/discordjs/discord.js/blob/1be67b88516b104073e46574a180498bad2aa02b/src/util/Util.js#L262-L263

File: test/Bitfield.test.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    expect(bf.any(Permissions.FLAGS.SEND_MESSAGES)).toStrictEqual(true);
+    expect(bf.any(Permissions.FLAGS.ADMINISTRATOR)).toStrictEqual(false);
+  });
+
+  test('Bitfield#equals', () => {
+    const bf = new Permissions([Permissions.FLAGS.SEND_MESSAGES, Permissions.FLAGS.VIEW_CHANNEL]);
+    // Redundant?
+    expect(
+      bf.equals(new Permissions([Permissions.FLAGS.SEND_MESSAGES, Permissions.FLAGS.VIEW_CHANNEL]).bitfield),
+    ).toStrictEqual(true);
```
Comment (Fyko) [inline test/Bitfield.test.js:20]: see comment

File: jest.config.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+	displayName: 'unit tests',
+	testRunner: "jest-circus/runner",
+	testEnvironment: 'node',
+	verbose: true,
+	testMatch: ['<rootDir>/test/*.test.js'],
+};
```
Comment (didinele) [inline jest.config.js:7]: Missing EOF new-line here

File: test/Bitfield.test.js (hunk 2 above)
Comment (didinele) [reply test/Bitfield.test.js:20]: Probably, but gotta get that test coverage up am I rite

**Review by didinele — CHANGES_REQUESTED** [review]
Other than that, tests LGTM.

Really only thing I can nit-pick is your over-use of `toStrictEqual`. Not necessary for primitives or things that you could generally compare, but I guess it's not that big of a deal.

File: .github/workflows/test-cron.yml [hunk 4]
```diff
@@ -77,3 +77,21 @@ jobs:
...
+
+      - name: Install Node v14
+        uses: actions/setup-node@v2
+        with:
+          node-version: 14
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Run TSLint
```
Comment (iCrawl) [inline .github/workflows/test-cron.yml:96]: TSLint?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -75,3 +75,21 @@ jobs:
...
+
+      - name: Install Node v14
+        uses: actions/setup-node@v2
+        with:
+          node-version: 14
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Run TSLint
```
Comment (iCrawl) [inline .github/workflows/test.yml:94]: TSLint?

File: package.json [hunk 6]
```diff
@@ -67,6 +68,7 @@
     "eslint-plugin-prettier": "^3.1.4",
     "husky": "^4.3.0",
     "jest": "^26.6.0",
+    "jest-circus": "^26.6.3",
```
Comment (iCrawl) [inline package.json:71]: Do you use the jest-circus runner for something specific?

File: test/Bitfield.test.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
+/* eslint-disable strict */
+const { Permissions } = require('../src');
+// You see, i understand this is supposed to be tests
```
Comment (iCrawl) [inline test/Bitfield.test.js:3]: That comment isn't really helpful at all.

File: test/constants.js [hunk 8]
```diff
@@ -0,0 +1 @@
...j3mgVHdBTp1AAoteFUa0cLhaSHhADMLynrSjDa9Pi3v8/jZuvuf776bb6NcVlCc/0S05BfjnEnUGSZd24GoWIjg5qacpcIopoWKnC1+ZzYx7Vfea9aeZF9CQSHiK2hzz2uJD3pZt8dJk0nSnMfjNSKtNQ+jD0+EkxgB/DyjKg+NdxVeWL0b/vXzmzSZ/gffotZkSd8coTodTox4cJuhMU14I5irXhJzpGfmNO9r5qRu39R1N/8z018bmmCuZjdnMsDhJUHWBVC6BuYRxfa2iU7swslX52Pl/la9XPU6EwVWk0gWcnWFPa0hrWkJx09JzS+D7yDTVJNeiugZiK7bL94mam9cqhWtaqiQyZgpyayTUSepS5YL08NPENS7WWB9DGfgeynHayKkX8EOoDXxSxaL8KIKg2kD7adnCX+pQRwH2XRDc9UcRWYO4e0YPMs0iMyxxMoMEKOEdO9ax1bzf2MemglxzLy0tuZizd0Npnm7TNs0c5QS6ci7Ua1WTGN3rh6r1ownpwmNmuNCSTjDRkM5VpftVquxOJRfNk4J6FWwtiUvHfjKmbMKiNeZAhhlxHXK47e8Xy2N9vTO2A2xyWWUGKWjLFdMP+zmpb5PxVogsyDkFmIcFatp8bacXFTQC8PgB5yufRlMiFAEiAmFVOCEKNIsm/vckhJCpSo7iRc1b+N343XSlWQftBRjDTXWQlOM2o2GNO/wzCE0DEj9CbKxCOjgexIAr/AS1ARMYDz4qPGKNGRKNnYNBzXLR2rznkWr33qmbPMy+gGb8aqHsHJOFCAfJtgBhyMpvkrBKTr/c7a+C1CgtLIrwrGKdqS8PIA+ML0vLTk/d5Z5kVGqjCKUDsKRwDykvB0ohKWYbZPJ7V1O3YB2Bth47eg/CVY1pe2ZUeBpSBLYdmzwOPAnLTCIWvGYVYiYvzECRJCATXfRXkH8NReHk4/zTJxzsRLbvlMx4IfSgbFU1Pea1dzAD8cEO4YMixCKGG4E0MJQRF9B+KHa8aJQXAIDyNsq4nWrRjaEJZh+OdthkEHwd72PYJSQTFTW+T9xZLoazHV+lVE05EDj9F+dAzAaU7HAJzmdAzAaU7HAJzmdAzAaU7HAJzmdAzAaU7/H2oTndI/U50mAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAABJRU5ErkJggg==";
```
Comment (iCrawl) [inline test/constants.js:1]: EOF New line

File: package.json (hunk 6 above)
Comment (Fyko) [reply package.json:71]: No real reason, just code transplanted from another repository.

**Review by catcosmicice — APPROVED** [review]

File: test/Bitfield.test.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    const bf = new Permissions([Permissions.FLAGS.SEND_MESSAGES, Permissions.FLAGS.VIEW_CHANNEL]);
+    expect(bf.any(Permissions.FLAGS.SEND_MESSAGES)).toStrictEqual(true);
+    expect(bf.any(Permissions.FLAGS.ADMINISTRATOR)).toStrictEqual(false);
+  });
+
+  test('Bitfield#equals', () => {
+    const bf = new Permissions([Permissions.FLAGS.SEND_MESSAGES, Permissions.FLAGS.VIEW_CHANNEL]);
+    // Redundant?
+    expect(
+      bf.equals(new Permissions([Permissions.FLAGS.SEND_MESSAGES, Permissions.FLAGS.VIEW_CHANNEL]).bitfield),
```
Comment (vladfrangu) [inline test/Bitfield.test.js:15]:
```suggestion
      bf.equals(bf.bitfield),
```

File: test/Util.test.js [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+      bar.set('a', 1);
+      bar.set('b', 2);
+      bar.set('c', 3);
+
+      const obj = {
+        foo: 'foo',
+        bar,
+      };
+
+      expect(Util.flatten(obj)).toStrictEqual({ foo: 'foo', bar: ['a', 'b', 'c'] });
```
Comment (vladfrangu) [inline test/Util.test.js:17]: This feels...wrong... why is bar an array of keys, and not an array of kv?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: test/Util.test.js (hunk 10 above)
Comment (Fyko) [reply test/Util.test.js:17]: because that is what Util#flatten does

Comment (vladfrangu) [reply test/Util.test.js:17]: But is that really what we want it to do? CC @iCrawl @kyranet

Comment (iCrawl) [reply test/Util.test.js:17]: I don't understand your question. Please elaborate.

Comment (Fyko) [reply test/Util.test.js:17]: I think they're suggesting that it it should flatten Collections into what would be `[key, value][]`

**Comment by twocountriestrus** [comment]
Gracias 


Enviado desde Yahoo Mail para iPhone


El viernes, abril 23, 2021, 10:35 a. m., Carter ***@***.***> escribió:




@Fyko commented on this pull request.

In test/Util.test.js:
> +const { Util, Collection } = require('../src');
+
+describe('Util unit tests', () => {
+  describe('Util#flatten', () => {
+    test('test collection flattening', () => {
+      const bar = new Collection();
+      bar.set('a', 1);
+      bar.set('b', 2);
+      bar.set('c', 3);
+
+      const obj = {
+        foo: 'foo',
+        bar,
+      };
+
+      expect(Util.flatten(obj)).toStrictEqual({ foo: 'foo', bar: ['a', 'b', 'c'] });

I think they're suggesting that it it should flatten Collections into what would be [key, value][], similar to Array#entries

File: test/Util.test.js (hunk 10 above)
Comment (iCrawl) [reply test/Util.test.js:17]: How is that a concern of this PR?

**Review by catcosmicice — APPROVED** [review]

File: test/Util.test.js (hunk 10 above)
Comment (kyranet) [reply test/Util.test.js:17]: IMO we should keep this PR `semver: patch` or `semver: N/A`, and not promote it to `semver: major`. A change in the utilities can lead to unexpected bugs in the library.

**Review by SpaceEEC — APPROVED** [review]

**Comment by iCrawl** [comment]
This needs a heavy rebase 👀

**Comment by vladfrangu** [comment]
This needs a rebase the size of a mountain

**Comment by iCrawl** [comment]
This has been stale for quite some time, if theres any interest in bringing it up to date, please reopen as a new PR.

**Pull request closed without merging**